pkg/kubelet/status/status_manager.go

```diff
@@ -265,7 +265,7 @@ func findContainerStatus(status *v1.PodStatus, containerID string) (containerSta
 
 }
 
-func (m *manager) TerminatePod(pod *v1.Pod) {
+func (m *manager) TerminatePod(pod *v1.Pod, podStatus *kubecontainer.PodStatus) {
```

Comment:
I don't know that you need to fetch new container status here because you're already going to be racing with a background status update (another thread can be retrieving status even as you're fetching it, so this pod status is already stale).

I would probably suggest trying to have your loop merge the status that is already on the pod - basically say "we aren't resetting the state, we're ensuring that all containers are marked as terminated it".

Pseudo code like:

```
for status.ContainerStatuses
  if containerIsAlreadyTerminatedAndNotRunning
    continue
  status.ContainerStatuses[I].State = Terminated{ exitCode == SAME_AS_SIGKILL }
```

In the linked issue, *all* containers are already stopped so the TerminatePod method doesn't have to set any status.  In fact, TerminatePod is called in one place, and if I read the logic right, the only place where containers should have to be terminated is when the pod is Failed or Succeeded (run once or evicted), in which case those are likely to get updated.

I think filling in "fake" container status is the real problem here.  We need an unambiguous signal from TerminatePod to the status manager that "this container status is not real, must be filled out, etc".  The Kubelet is *required* to update all status before it terminates, so it's not as if we don't have the info.  

I'm kind of questioning why TerminatePod was clearing status in the first place - I would be tempted to completely remove that and add additional tests that verify that async terminatePod along with detecting pod deletion is sufficient for the status manager to correctly drain.  We *have* to be able to handle the race conditions between terminatePod and syncPod, and whomever owns the job of deleting the pod is the one who needs to ensure that all status is flushed (and only if we are in a state where we know no more status can be gathered should we be defaulting status).